README.md
# PHP Date Parser

[![Release](https://img.shields.io/github/v/release/ixnode/php-date-parser)](https://github.com/ixnode/php-date-parser/releases)
[![](https://img.shields.io/github/release-date/ixnode/php-date-parser)](https://github.com/ixnode/php-date-parser/releases)
![](https://img.shields.io/github/repo-size/ixnode/php-date-parser.svg)
[![PHP](https://img.shields.io/badge/PHP-^8.2-777bb3.svg?logo=php&logoColor=white&labelColor=555555&style=flat)](https://www.php.net/supported-versions.php)
[![PHPStan](https://img.shields.io/badge/PHPStan-Level%20Max-777bb3.svg?style=flat)](https://phpstan.org/user-guide/rule-levels)
[![PHPUnit](https://img.shields.io/badge/PHPUnit-Unit%20Tests-6b9bd2.svg?style=flat)](https://phpunit.de)
[![PHPCS](https://img.shields.io/badge/PHPCS-PSR12-416d4e.svg?style=flat)](https://www.php-fig.org/psr/psr-12/)
[![PHPMD](https://img.shields.io/badge/PHPMD-ALL-364a83.svg?style=flat)](https://github.com/phpmd/phpmd)
[![Rector - Instant Upgrades and Automated Refactoring](https://img.shields.io/badge/Rector-PHP%208.2-73a165.svg?style=flat)](https://github.com/rectorphp/rector)
[![LICENSE](https://img.shields.io/github/license/ixnode/php-api-version-bundle)](https://github.com/ixnode/php-api-version-bundle/blob/master/LICENSE)

> This library parses various given date and time strings into
> DateTime or DateTimeImmutable classes which return the time
> range. Can be used e.g. excellently for command line
> arguments and options to make database queries with.

## Examples / Usage

```php
use Ixnode\PhpDateParser\DateParser;
```

### Date parser (`UTC`)

```php
print (new DateParser('2023-07-01'))->formatFrom('Y-m-d H:i:s');
// 2023-07-01 00:00:00

print (new DateParser('2023-07-01'))->formatTo('Y-m-d H:i:s');
// 2023-07-01 23:59:59
```

### Word parser (`UTC`)

*  Imagine that now is the time: `2023-07-07 12:34:56`

```php
print (new DateParser('today'))->formatFrom('Y-m-d H:i:s');
// 2023-07-07 00:00:00

print (new DateParser('today'))->formatTo('Y-m-d H:i:s');
// 2023-07-07 23:59:59
```

### Date parser with timezones

* Input: `America/New_York`
* Output: `Europe/Berlin`

```php
/* Parses given date time from timezone America/New_York; Output to timezone Europe/Berlin */
print (new DateParser('<2023-07-01', 'America/New_York'))->formatFrom('Y-m-d H:i:s', 'Europe/Berlin');
// null

/* Parses given date time from timezone America/New_York; Output to timezone Europe/Berlin */
print (new DateParser('<2023-07-01', 'America/New_York'))->formatTo('Y-m-d H:i:s', 'Europe/Berlin');
// 2023-07-01 05:59:59
```

### Working with `DateRange` class

```php
/* Parses given date time from timezone America/New_York */
$dateParser = (new DateParser('2023-07-01', 'America/New_York'));

/* Sets default output to timezone Asia/Tokyo */
$dateRange = $dateParser->getDateRange('Asia/Tokyo');

print $dateRange->getFrom()?->format('Y-m-d H:i:s (e)');
// 2023-07-01 13:00:00 (Asia/Tokyo)

print $dateRange->getTo()?->format('Y-m-d H:i:s (e)');
// 2023-07-02 12:59:59 (Asia/Tokyo)
```

## Parsing formats

### Supported words

| Word          | Description                   |
|---------------|-------------------------------|
| `next-second` | Next second (`'s' + 1`)       |
| `this-second` | This second (`'s'`)           |
| `last-second` | Last second (`'s' - 1`)       |
| `next-minute` | Next minute (`'i' + 1`)       |
| `this-minute` | This minute (`'i'`)           |
| `last-minute` | Last minute (`'i' - 1`)       |
| `next-hour`   | Next hour (`'G' + 1`)         |
| `this-hour`   | This hour (`'G'`)             |
| `last-hour`   | Last hour (`'G' - 1`)         |
| `tomorrow`    | The day tomorrow (`'j' + 1`)  |
| `today`       | The day today (`'j'`)         |
| `yesterday`   | The day yesterday (`'j' - 1`) |
| `next-month`  | Next month (`'n' + 1`)        |
| `this-month`  | This month (`'n'`)            |
| `last-month`  | Last month (`'n' - 1`)        |
| `next-year`   | Next year (`'Y' + 1`)         |
| `this-year`   | This year (`'Y'`)             |
| `last-year`   | Last year (`'Y' - 1`)         |

### Overview

#### Exact time parser (`=datetime`)

* Imagine that now is the time: `2023-07-07 12:34:56`

| Given format                 | Description                                                                            | From `('Y-m-d H:i:s')`               | To `('Y-m-d H:i:s')`                 |
|------------------------------|----------------------------------------------------------------------------------------|--------------------------------------|--------------------------------------|
| <nobr>`"next-hour"`</nobr>   | Returns the date range from the next hour. Alias of `"2023-07-07 13"`.                 | <nobr>`"2023-07-07 13:00:00"`</nobr> | <nobr>`"2023-07-07 13:59:59"`</nobr> |
| <nobr>`"=next-hour"`</nobr>  | Alias of `"next-hour"`.                                                                | <nobr>`"2023-07-07 13:00:00"`</nobr> | <nobr>`"2023-07-07 13:59:59"`</nobr> |
| <nobr>`"this-hour"`</nobr>   | Returns the date range from current hour. Alias of `"2023-07-07 12"`.                  | <nobr>`"2023-07-07 12:00:00"`</nobr> | <nobr>`"2023-07-07 12:59:59"`</nobr> |
| <nobr>`"=this-hour"`</nobr>  | Alias of `"this-hour"`.                                                                | <nobr>`"2023-07-07 12:00:00"`</nobr> | <nobr>`"2023-07-07 12:59:59"`</nobr> |
| <nobr>`"last-hour"`</nobr>   | Returns the date range from the last hour. Alias of `"2023-07-07 11"`.                 | <nobr>`"2023-07-07 11:00:00"`</nobr> | <nobr>`"2023-07-07 11:59:59"`</nobr> |
| <nobr>`"=last-hour"`</nobr>  | Alias of `"last-hour"`.                                                                | <nobr>`"2023-07-07 11:00:00"`</nobr> | <nobr>`"2023-07-07 11:59:59"`</nobr> |
| <nobr>`"tomorrow"`</nobr>    | Returns the date range from tomorrow. Alias of `"2023-07-08"`.                         | <nobr>`"2023-07-08 00:00:00"`</nobr> | <nobr>`"2023-07-08 23:59:59"`</nobr> |
| <nobr>`"=tomorrow"`</nobr>   | Alias of `"tomorrow"`.                                                                 | <nobr>`"2023-07-08 00:00:00"`</nobr> | <nobr>`"2023-07-08 23:59:59"`</nobr> |
| <nobr>`"today"`</nobr>       | Returns the date range from today. Alias of `"2023-07-07"`.                            | <nobr>`"2023-07-07 00:00:00"`</nobr> | <nobr>`"2023-07-07 23:59:59"`</nobr> |
| <nobr>`"=today"`</nobr>      | Alias of `"today"`.                                                                    | <nobr>`"2023-07-07 00:00:00"`</nobr> | <nobr>`"2023-07-07 23:59:59"`</nobr> |
| <nobr>`"yesterday"`</nobr>   | Returns the date range from yesterday. Alias of `"2023-07-06"`.                        | <nobr>`"2023-07-06 00:00:00"`</nobr> | <nobr>`"2023-07-06 23:59:59"`</nobr> |
| <nobr>`"=yesterday"`</nobr>  | Alias of `"yesterday"`                                                                 | <nobr>`"2023-07-06 00:00:00"`</nobr> | <nobr>`"2023-07-06 23:59:59"`</nobr> |
| <nobr>`"this-month"`</nobr>  | Date range from first day to last day this month. Alias of `"2023-07-01\|2023-07-31"`. | <nobr>`"2023-07-01 00:00:00"`</nobr> | <nobr>`"2023-07-31 23:59:59"`</nobr> |
| <nobr>`"=this-month"`</nobr> | Alias of `"this-month"`                                                                | <nobr>`"2023-07-01 00:00:00"`</nobr> | <nobr>`"2023-07-31 23:59:59"`</nobr> |
| <nobr>`"2023-07-01"`</nobr>  | Exactly the given date.                                                                | <nobr>`"2023-07-01 00:00:00"`</nobr> | <nobr>`"2023-07-01 23:59:59"`</nobr> |
| <nobr>`"=2023-07-01"`</nobr> | Alias of `"2023-07-01"`                                                                | <nobr>`"2023-07-01 00:00:00"`</nobr> | <nobr>`"2023-07-01 23:59:59"`</nobr> |

#### Time is greater than parser (`>from`)

* Imagine that now is the time: `2023-07-07 12:34:56`
* "To" values are `NULL`

| Given format                            | Description                                             | From `('Y-m-d H:i:s')`               | To `('Y-m-d H:i:s')`                 |
|-----------------------------------------|---------------------------------------------------------|--------------------------------------|--------------------------------------|
| <nobr>`">tomorrow"`</nobr>              | Later than tomorrow<sup>1)</sup>                        | <nobr>`"2023-07-09 00:00:00"`</nobr> | <nobr>`NULL`</nobr>                  |
| <nobr>`">=tomorrow"`</nobr>             | Later than tomorrow<sup>2)</sup>                        | <nobr>`"2023-07-08 00:00:00"`</nobr> | <nobr>`NULL`</nobr>                  |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Releases

### [1.0.3] - 2023-07-22

* README.md updates

### [1.0.2] - 2023-07-22

* Add duration method to DateRange and DateParser class
* Throw exception if "to" value is lower than "from" value

### [1.0.1] - 2023-07-22

* Add badges to README.md

### [1.0.0] - 2023-07-22

* Switch to version 1.0.0

### [0.1.13] - 2023-07-22

* README.md updates

### [0.1.12] - 2023-07-22

* README.md updates

### [0.1.11] - 2023-07-22

* Refactoring
* README.md updates

### [0.1.10] - 2023-07-21

* Add autoload directory within vendor

### [0.1.9] - 2023-07-20

* Add command line interpreter
* Fix recursive issue with output timezone
* Update README.md

### [0.1.8] - 2023-07-18

* Update README.md
* Add more DateRange tests

### [0.1.7] - 2023-07-17

* Add hour, minute and second parser
* Add this, next, last parser for hour, minute and second

### [0.1.6] - 2023-07-16

* Add timezone support

### [0.1.5] - 2023-07-09

* Add more parser options like next-year, last-year, etc.

### [0.1.4] - 2023-07-08

* Add more parser options like next-month, this-month, last-month, this-year, etc.
* README.md updates

### [0.1.3] - 2023-07-08

* Add more parser options like tomorrow, etc.
* README.md updates
* Refactoring

### [0.1.2] - 2023-07-08

* Add "yesterday" to "today" to README.md and tests.

### [0.1.1] - 2023-07-08

* README.md changes


### [0.1.0] - 2023-07-07

* Initial release with the first implementation
* Add src
* Add tests
  * PHP Coding Standards Fixer
  * PHPMND - PHP Magic Number Detector
  * PHPStan - PHP Static Analysis Tool
  * PHPUnit - The PHP Testing Framework
  * Rector - Instant Upgrades and Automated Refactoring
* Add README.md
* Add LICENSE.md

## Add new version

```bash
# Checkout master branch
$ git checkout main && git pull

# Check current version
$ vendor/bin/version-manager --current

# Increase patch version
$ vendor/bin/version-manager --patch

# Change changelog
$ vi CHANGELOG.md

# Push new version
$ git add CHANGELOG.md VERSION && git commit -m "Add version $(cat VERSION)" && git push

# Tag and push new version
$ git tag -a "$(cat VERSION)" -m "Version $(cat VERSION)" && git push origin "$(cat VERSION)"
```


src/Base/BaseDateParser.php
<?php

/*
 * This file is part of the ixnode/php-date-parser project.
 *
 * (c) Björn Hempel <https://www.hempel.li/>
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

declare(strict_types=1);

namespace Ixnode\PhpDateParser\Base;

use DateTime;
use DateTimeImmutable;
use DateTimeZone;
use Exception;
use Ixnode\PhpDateParser\Constants\Timezones;
use Ixnode\PhpDateParser\DateRange;
use Ixnode\PhpException\Case\CaseUnsupportedException;
use Ixnode\PhpException\Parser\ParserException;
use Ixnode\PhpException\Type\TypeInvalidException;

/**
 * Class BaseDateParser
 *
 * @author Björn Hempel <bjoern@hempel.li>
 * @version 0.1.0 (2023-07-08)
 * @since 0.1.0 (2023-07-08) First version.
 * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
 */
class BaseDateParser
{
    final public const HOUR_FIRST = 0;

    final public const MINUTE_FIRST = 0;

    final public const SECOND_FIRST = 0;

    final public const HOUR_LAST = 23;

    final public const MINUTE_LAST = 59;

    final public const SECOND_LAST = 59;

    final public const FORMAT_DEFAULT = 'Y-m-d H:i:s';

    final public const FORMAT_DATE = 'Y-m-d';

    final public const FORMAT_DATE_REGEXP = '~^([0-9]{4})-([0-9]{2})-([0-9]{2})$~';

    final public const FORMAT_DATE_HOUR = 'Y-m-d H';

    final public const FORMAT_DATE_HOUR_REGEXP = '~^([0-9]{4})-([0-9]{2})-([0-9]{2}) ([0-9]{2})$~';

    final public const FORMAT_DATE_HOUR_MINUTE = 'Y-m-d H:i';

    final public const FORMAT_DATE_HOUR_MINUTE_REGEXP = '~^([0-9]{4})-([0-9]{2})-([0-9]{2}) ([0-9]{2}):([0-9]{2})$~';

    final public const FORMAT_DATE_HOUR_MINUTE_SECOND = 'Y-m-d H:i:s';

    final public const FORMAT_DATE_HOUR_MINUTE_SECOND_REGEXP = '~^([0-9]{4})-([0-9]{2})-([0-9]{2}) ([0-9]{2}):([0-9]{2}):([0-9]{2})$~';

    final public const FORMAT_THIS_MONTH_FIRST = 'Y-m-1';

    final public const FORMAT_THIS_MONTH_LAST = 'Y-m-t';

    final public const FORMAT_THIS_YEAR_FIRST = 'Y-1-1';

    final public const FORMAT_THIS_YEAR_LAST = 'Y-12-31';

    final public const VALUE_NOW = 'now';

    final public const VALUE_NEXT_SECOND = 'next-second';

    final public const VALUE_THIS_SECOND = 'this-second';

    final public const VALUE_LAST_SECOND = 'last-second';

    final public const VALUE_NEXT_MINUTE = 'next-minute';

    final public const VALUE_THIS_MINUTE = 'this-minute';

    final public const VALUE_LAST_MINUTE = 'last-minute';

    final public const VALUE_NEXT_HOUR = 'next-hour';

    final public const VALUE_THIS_HOUR = 'this-hour';

    final public const VALUE_LAST_HOUR = 'last-hour';

    final public const VALUE_TOMORROW = 'tomorrow';

    final public const VALUE_TODAY = 'today';

    final public const VALUE_YESTERDAY = 'yesterday';

    final public const VALUE_NEXT_MONTH = 'next-month';

    final public const VALUE_THIS_MONTH = 'this-month';

    final public const VALUE_LAST_MONTH = 'last-month';

    final public const VALUE_NEXT_YEAR = 'next-year';

    final public const VALUE_THIS_YEAR = 'this-year';

    final public const VALUE_LAST_YEAR = 'last-year';

    final public const SECONDS_A_DAY = 86400;

    final public const SECONDS_AN_HOUR = 3600;

    final public const SECONDS_A_MINUTE = 60;

    final public const SECONDS_A_SECOND = 1;

    protected int $hourFirst = self::HOUR_FIRST;

    protected int $minuteFirst = self::MINUTE_FIRST;

    protected int $secondFirst = self::SECOND_FIRST;

    protected int $hourLast = self::HOUR_LAST;

    protected int $minuteLast = self::MINUTE_LAST;

    protected int $secondLast = self::SECOND_LAST;

    protected string|null $range;

    protected DateRange $dateRange;

    /**
     * @param string|null $range
     * @param DateTimeZone $dateTimeZoneInput
     * @throws ParserException
     * @throws TypeInvalidException
     * @throws CaseUnsupportedException
     * @throws Exception
     */
    public function __construct(string|null $range, protected DateTimeZone $dateTimeZoneInput = new DateTimeZone('UTC'))
    {
        $this->range = !is_null($range) ? trim(strtolower($range)) : null;

        $dateRange = $this->parseRange($range);

        $this->dateRange = $this->getDateRangeInstance(
            $dateRange->getFrom(),
            $dateRange->getTo(),
            $dateTimeZoneInput
        )->setDefaultDateTimeZone(new DateTimeZone(Timezones::UTC));
    }

    /**
     * Returns the default date time zone.
     *
     * @return DateTimeZone
     */
    public function getDefaultDateTimeZone(): DateTimeZone
    {
        return $this->dateRange->getDefaultDateTimeZone();
    }

    /**
     * Sets the default date time zone.
     *
     * @param DateTimeZone|string $defaultDateTimeZone
     * @return self
     * @throws Exception
     */
    public function setDefaultDateTimeZone(DateTimeZone|string $defaultDateTimeZone): self
    {
        $this->dateRange->setDefaultDateTimeZone($defaultDateTimeZone);

        return $this;
    }

    /**
     * Parses the given date range.
     *
     * @param string|null $range
     * @return DateRange
     * @throws TypeInvalidException
     * @throws ParserException
     * @throws Exception
     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
     */
    private function parseRange(string|null $range): DateRange
    {
        if (is_null($range)) {
            return $this->getDateRangeInstance(
                null,
                null
            );
        }

        $this->setTimes();

        $matches = [];
        switch (true) {

            /* ==================
             * Second word parser
             * ==================
             */
            case $range === self::VALUE_NEXT_SECOND:
            case $range === self::VALUE_THIS_SECOND:
            case $range === self::VALUE_LAST_SECOND:
                /* Parses "next-second", "this-second", "last-second" date time. */
                $dateTimeFrom = $this->getDateTimeRaw(self::VALUE_TODAY, $this->getThisHour(), $this->getThisMinute(), $this->getThisSecond());
                $dateTimeTo = $this->getDateTimeRaw(self::VALUE_TODAY, $this->getThisHour(), $this->getThisMinute(), $this->getThisSecond());
                $direction = $this->getDirection($range, self::VALUE_NEXT_SECOND, self::VALUE_LAST_SECOND);
                $this->modifyDateTime($dateTimeFrom, self::SECONDS_A_SECOND, $direction);
                $this->modifyDateTime($dateTimeTo, self::SECONDS_A_SECOND, $direction);
                return $this->getDateRangeInstance($dateTimeFrom, $dateTimeTo);



            /* ==================
             * Minute word parser
             * ==================
             */
            case $range === self::VALUE_NEXT_MINUTE:
            case $range === self::VALUE_THIS_MINUTE:
            case $range === self::VALUE_LAST_MINUTE:
                /* Parses "next-minute", "this-minute", "last-minute" date time. */
                $dateTimeFrom = $this->getDateTimeRaw(self::VALUE_TODAY, $this->getThisHour(), $this->getThisMinute(), $this->secondFirst);
                $dateTimeTo = $this->getDateTimeRaw(self::VALUE_TODAY, $this->getThisHour(), $this->getThisMinute(), $this->secondLast);
                $direction = $this->getDirection($range, self::VALUE_NEXT_MINUTE, self::VALUE_LAST_MINUTE);
                $this->modifyDateTime($dateTimeFrom, self::SECONDS_A_MINUTE, $direction);
                $this->modifyDateTime($dateTimeTo, self::SECONDS_A_MINUTE, $direction);
                return $this->getDateRangeInstance($dateTimeFrom, $dateTimeTo);



            /* ================
             * Hour word parser
             * ================
             */
            case $range === self::VALUE_NEXT_HOUR:
            case $range === self::VALUE_THIS_HOUR:
            case $range === self::VALUE_LAST_HOUR:
                /* Parses "next-hour", "this-hour", "last-hour" date time. */
                $dateTimeFrom = $this->getDateTimeRaw(self::VALUE_TODAY, $this->getThisHour(), $this->minuteFirst, $this->secondFirst);
                $dateTimeTo = $this->getDateTimeRaw(self::VALUE_TODAY, $this->getThisHour(), $this->minuteLast, $this->secondLast);
                $direction = $this->getDirection($range, self::VALUE_NEXT_HOUR, self::VALUE_LAST_HOUR);
                $this->modifyDateTime($dateTimeFrom, self::SECONDS_AN_HOUR, $direction);
                $this->modifyDateTime($dateTimeTo, self::SECONDS_AN_HOUR, $direction);
                return $this->getDateRangeInstance($dateTimeFrom, $dateTimeTo);



            /* ===============
             * Day word parser
             * ===============
             */
            case $range === self::VALUE_TOMORROW:
                /* Parses "tomorrow" date. */
                return $this->getDateRangeInstance(
                    $this->getDateTimeRaw(self::VALUE_TOMORROW, $this->hourFirst, $this->minuteFirst, $this->secondFirst),
                    $this->getDateTimeRaw(self::VALUE_TOMORROW, $this->hourLast, $this->minuteLast, $this->secondLast)
                );
            case $range === self::VALUE_TODAY:
                /* Parses "today" date. */
                return $this->getDateRangeInstance(
                    $this->getDateTimeRaw(self::VALUE_NOW, $this->hourFirst, $this->minuteFirst, $this->secondFirst),
                    $this->getDateTimeRaw(self::VALUE_NOW, $this->hourLast, $this->minuteLast, $this->secondLast)
                );
            case $range === self::VALUE_YESTERDAY:
                /* Parses "yesterday" date. */
                return $this->getDateRangeInstance(
                    $this->getDateTimeRaw(self::VALUE_YESTERDAY, $this->hourFirst, $this->minuteFirst, $this->secondFirst),
                    $this->getDateTimeRaw(self::VALUE_YESTERDAY, $this->hourLast, $this->minuteLast, $this->secondLast)
                );



            /* =================
             * Month word parser
             * =================
             */
            case $range === self::VALUE_NEXT_MONTH:
                /* Parses "next-month" date. */
                return $this->getDateRangeNextMonth();
            case $range === self::VALUE_THIS_MONTH:
                /* Parses "this-month" date. */
                return $this->getDateRangeInstance(
                    $this->getDateTimeRaw(date(self::FORMAT_THIS_MONTH_FIRST), $this->hourFirst, $this->minuteFirst, $this->secondFirst),
                    $this->getDateTimeRaw(date(self::FORMAT_THIS_MONTH_LAST), $this->hourLast, $this->minuteLast, $this->secondLast)
                );
            case $range === self::VALUE_LAST_MONTH:
                /* Parses "last-month" date. */
                return $this->getDateRangeLastMonth();



            /* ================
             * Year word parser
             * ================
             */
            case $range === self::VALUE_NEXT_YEAR:
                /* Parses "next-year" date. */
                return $this->getDateRangeNextYear();
            case $range === self::VALUE_THIS_YEAR:
                /* Parses "this-year" date. */
                return $this->getDateRangeInstance(
                    $this->getDateTimeRaw(date(self::FORMAT_THIS_YEAR_FIRST), $this->hourFirst, $this->minuteFirst, $this->secondFirst),
                    $this->getDateTimeRaw(date(self::FORMAT_THIS_YEAR_LAST), $this->hourLast, $this->minuteLast, $this->secondLast)
                );
            case $range === self::VALUE_LAST_YEAR:
                /* Parses "last-year" date. */
                return $this->getDateRangeLastYear();



            /* =======================
             * Earlier than parser (<)
             * =======================
             */
            case preg_match('~^(<[+=]|-)~', $range, $matches) === 1:
                /* Starts with <+: parses a "∞ (infinity)" to given "from" date (including given date). */
                return $this->getDateRangeInstance(
                    null,
                    $this->parseRange(substr($range, strlen($matches[1])))->getTo()
                );
            case str_starts_with($range, '<'):
                /* Starts with <: parses a "∞ (infinity)" to given "from" date (excluding given date). */
                return $this->getDateRangeInstance(
                    null,
                    $this->parseRange(substr($range, 1))->getFrom()?->modify('-1 second')
                );



            /* =======================
             * Earlier than parser (<)
             * =======================
             */
            case preg_match('~^(>[+=]|[+])~', $range, $matches) === 1:
                /* Starts with >+: parses a given "from" (including given date) to "∞ (infinity)" date. */
                return $this->getDateRangeInstance(
                    $this->parseRange(substr($range, strlen($matches[1])))->getFrom(),
                    null
                );
            case str_starts_with($range, '>'):
                /* Starts with >: parses a given "from" (excluding given date) to "∞ (infinity)" date. */
                return $this->getDateRangeInstance(
                    $this->parseRange(substr($range, 1))->getTo()?->modify('+1 second'),
                    null
                );



            /* ================
             * Range parser (|)
             * ================
             */
            case str_contains($range, '|'):
                /* Starts with |: parses a given "from" (including given date) to "to" date (including given date). */
                $splitted = explode('|', $range);
                return $this->getDateRangeInstance(
                    $this->parseRange($splitted[0])->getFrom(),
                    $this->parseRange($splitted[1])->getTo()
                );



            /* ================
             * Equal parser (=)
             * ================
             */
            case str_starts_with($range, '='):
                /* Starts with =: a given date exactly. */
                return $this->getDateRangeInstance(
                    $this->parseRange(substr($range, 1))->getFrom(),
                    $this->parseRange(substr($range, 1))->getTo(),
                );



            /* ================
             * Date parser (Y-m-d [H[:i[:s]]])
             * ================
             */
            default:
                /* Parse the date */
                $dateTime = $this->parseDateTime($range);
                return $this->getDateRangeInstance(
                    (clone $dateTime)->setTime($this->hourFirst, $this->minuteFirst, $this->secondFirst),
                    (clone $dateTime)->setTime($this->hourLast, $this->minuteLast, $this->secondLast),
                );
        }
    }

    /**
     * Returns the DateRange Object from next month.
     *
     * @return DateRange
     * @throws Exception
     */
    private function getDateRangeNextMonth(): DateRange
    {
        $firstNextMonth = (new DateTime(date(self::FORMAT_THIS_MONTH_LAST)))
            ->setTime($this->hourLast, $this->minuteLast, $this->secondLast)
            ->modify('+1 second');

        $lastNextMonth = (new DateTime($firstNextMonth->format(self::FORMAT_THIS_MONTH_LAST)))
            ->setTime($this->hourLast, $this->minuteLast, $this->secondLast);

        return $this->getDateRangeInstance($firstNextMonth, $lastNextMonth);
    }

    /**
     * Returns the DateRange Object from last month.
     *
     * @return DateRange
     * @throws Exception
     */
    private function getDateRangeLastMonth(): DateRange
    {
        $lastLastMonth = (new DateTime(date(self::FORMAT_THIS_MONTH_FIRST)))
            ->setTime($this->hourFirst, $this->minuteFirst, $this->secondFirst)
            ->modify('-1 second')
            ->setTime($this->hourLast, $this->minuteLast, $this->secondLast);

        $firstLastMonth = (new DateTime($lastLastMonth->format(self::FORMAT_THIS_MONTH_FIRST)))
            ->setTime($this->hourFirst, $this->minuteFirst, $this->secondFirst);

        return $this->getDateRangeInstance($firstLastMonth, $lastLastMonth);
    }

    /**
     * Returns the DateRange Object from next year.
     *
     * @return DateRange
     * @throws Exception
     */
    private function getDateRangeNextYear(): DateRange
    {
        $firstNextYear = (new DateTime(date(self::FORMAT_THIS_YEAR_LAST)))
            ->setTime($this->hourLast, $this->minuteLast, $this->secondLast)
            ->modify('+1 second');

        $lastNextYear = (new DateTime($firstNextYear->format(self::FORMAT_THIS_YEAR_LAST)))
            ->setTime($this->hourLast, $this->minuteLast, $this->secondLast);

        return $this->getDateRangeInstance($firstNextYear, $lastNextYear);
    }

    /**
     * Returns the DateRange Object from last year.
     *
     * @return DateRange
     * @throws Exception
     */
    private function getDateRangeLastYear(): DateRange
    {
        $lastLastYear = (new DateTime(date(self::FORMAT_THIS_YEAR_FIRST)))
            ->setTime($this->hourFirst, $this->minuteFirst, $this->secondFirst)
            ->modify('-1 second')
            ->setTime($this->hourLast, $this->minuteLast, $this->secondLast);

        $firstLastYear = (new DateTime($lastLastYear->format(self::FORMAT_THIS_YEAR_FIRST)))
            ->setTime($this->hourFirst, $this->minuteFirst, $this->secondFirst);

        return $this->getDateRangeInstance($firstLastYear, $lastLastYear);
    }

    /**
     * Sets first and last times.
     *
     * @param int|null $hourFirst
     * @param int|null $minuteFirst
     * @param int|null $secondFirst
     * @param int|null $hourLast
     * @param int|null $minuteLast
     * @param int|null $secondLast
     * @return void
     */
    private function setTimes(
        int|null $hourFirst = null,
        int|null $minuteFirst = null,
        int|null $secondFirst = null,
        int|null $hourLast = null,
        int|null $minuteLast = null,
        int|null $secondLast = null
    ): void
    {
        $this->hourFirst = $hourFirst ?: self::HOUR_FIRST;
        $this->minuteFirst = $minuteFirst ?: self::MINUTE_FIRST;
        $this->secondFirst = $secondFirst ?: self::SECOND_FIRST;
        $this->hourLast = $hourLast ?: self::HOUR_LAST;
        $this->minuteLast = $minuteLast ?: self::MINUTE_LAST;
        $this->secondLast = $secondLast ?: self::SECOND_LAST;
    }

    /**
     * Parses the given date time.
     *
     * @param string $dateTime
     * @return DateTime
     * @throws TypeInvalidException
     * @throws ParserException
     */
    protected function parseDateTime(string $dateTime): DateTime
    {
        $matches = [];
        switch (true) {
            case preg_match(self::FORMAT_DATE_HOUR_MINUTE_SECOND_REGEXP, $dateTime, $matches) > 0:
                $dateTime = date_create_from_format(self::FORMAT_DATE_HOUR_MINUTE_SECOND, $dateTime);
                $this->setTimes(
                    (int) $matches[4],
                    (int) $matches[5],
                    (int) $matches[6],
                    (int) $matches[4],
                    (int) $matches[5],
                    (int) $matches[6]
                );
                break;

            case preg_match(self::FORMAT_DATE_HOUR_MINUTE_REGEXP, $dateTime, $matches) > 0:
                $dateTime = date_create_from_format(self::FORMAT_DATE_HOUR_MINUTE, $dateTime);
                $this->setTimes(
                    (int) $matches[4],
                    (int) $matches[5],
                    null,
                    (int) $matches[4],
                    (int) $matches[5]
                );
                break;

            case preg_match(self::FORMAT_DATE_HOUR_REGEXP, $dateTime, $matches) > 0:
                $dateTime = date_create_from_format(self::FORMAT_DATE_HOUR, $dateTime);
                $this->setTimes(
                    (int) $matches[4],
                    null,
                    null,
                    (int) $matches[4]
                );
                break;

            case preg_match(self::FORMAT_DATE_REGEXP, $dateTime, $matches) > 0:
                $dateTime = date_create_from_format(self::FORMAT_DATE, $dateTime);
                $this->setTimes();
                break;

            default:
                throw new ParserException($dateTime, sprintf(
                 'String must be in the format "%s", "%s", "%s" or "%s".',
                 self::FORMAT_DATE,
                 self::FORMAT_DATE_HOUR,
                 self::FORMAT_DATE_HOUR_MINUTE,
                 self::FORMAT_DATE_HOUR_MINUTE_SECOND
               ));
        }

        if (!$dateTime instanceof DateTime) {
            throw new TypeInvalidException('object', gettype($dateTime));
        }

        return $dateTime;
    }

    /**
     * @param string $dateTime
     * @param int $hour
     * @param int $minute
     * @param int $second
     * @return DateTime
     * @throws Exception
     */
    protected function getDateTimeRaw(string $dateTime, int $hour, int $minute, int $second): DateTime
    {
        return (new DateTime($dateTime))->setTime($hour, $minute, $second);
    }

    /**
     * Returns the DateRange instance from given date range.
     *
     * @param DateTime|DateTimeImmutable|null $from
     * @param DateTime|DateTimeImmutable|null $to
     * @param DateTimeZone $dateTimeZoneInput
     * @return DateRange
     * @throws CaseUnsupportedException
     * @SuppressWarnings(PHPMD.ShortVariable)
     */
    protected function getDateRangeInstance(
        DateTime|DateTimeImmutable|null $from,
        DateTime|DateTimeImmutable|null $to,
        DateTimeZone $dateTimeZoneInput = new DateTimeZone(Timezones::UTC)
    ): DateRange
    {
        return new DateRange($from, $to, $dateTimeZoneInput);
    }

    /**
     * Returns this hour.
     *
     * @return int
     */
    private function getThisHour(): int
    {
        return (int) date('H');
    }

    /**
     * Returns this minute.
     *
     * @return int
     */
    private function getThisMinute(): int
    {
        return (int) date('i');
    }

    /**
     * Returns this second.
     *
     * @return int
     */
    private function getThisSecond(): int
    {
        return (int) date('s');
    }

    /**
     * Modify given date time object by given seconds.
     *
     * @param DateTime $dateTime
     * @param bool|null $direction
     * @param int $seconds
     * @return void
     */
    private function modifyDateTime(DateTime $dateTime, int $seconds, bool|null $direction): void
    {
        if ($direction === null) {
            return;
        }

        $dateTime->modify(sprintf('%s%d seconds', $direction ? '+' : '-', $seconds));
    }

    /**
     * Returns the direction of given range (next, last, etc.)
     *
     * @param string $range
     * @param string $next
     * @param string $last
     * @return bool|null
     */
    private function getDirection(string $range, string $next, string $last): bool|null
    {
        return match ($range) {
            $next => true,
            $last => false,
            default => null,
        };
    }
}
